core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpFunctions.java

```diff
@@ -0,0 +1,117 @@
+import com.linecorp.armeria.common.http.HttpResponse;
+import com.linecorp.armeria.common.util.Types;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+/**
+ * Provides various utility functions for internal use related with {@link DynamicHttpFunction}.
+ */
+final class DynamicHttpFunctions {
+
+    /**
+     * Returns {@link ResponseConverter} which can convert given {@code type} into {@link HttpResponse}
+     * from the configured converters.
+     *
+     * @throws IllegalArgumentException if no appropriate {@link ResponseConverter} exists for given
+     * {@code type}
+     */
+    private static ResponseConverter converter(Class<?> type, Map<Class<?>, ResponseConverter> converters) {
+        // search for the converter mapped to itself or one of its superclasses, except Object.class.
+        Class<?> current = type;
+        while (current != Object.class) {
+            ResponseConverter converter = converters.get(current);
+            if (converter != null) {
+                return converter;
+            }
+            current = current.getSuperclass();
+        }
+
+        // search for the converter mapped to one of its interface.
+        for (Class<?> itfc : Types.getAllInterfaces(type)) {
+            ResponseConverter converter = converters.get(itfc);
+            if (converter != null) {
+                return converter;
+            }
+        }
+
+        // search for the converter mapped to Object.class.
+        if (converters.containsKey(Object.class)) {
+            return converters.get(Object.class);
+        }
+
+        // No appropriate converter found: raise runtime exception.
+        throw new IllegalArgumentException(
+                "No available converter exists for " + type.getSimpleName() + " type");
+    }
+
+    /**
+     * Converts {@code object} into {@link HttpResponse}, using one of given {@code converters}.
+     *
+     * @throws IllegalStateException if an {@link Exception} thrown during conversion
+     */
+    private static HttpResponse convert(Object object, Map<Class<?>, ResponseConverter> converters) {
+        if (HttpResponse.class.isAssignableFrom(object.getClass())) {
+            return (HttpResponse) object;
+        } else {
+            ResponseConverter converter = converter(object.getClass(), converters);
+            try {
+                return converter.convert(object);
+            } catch (Exception e) {
+                throw new IllegalStateException("Exception occurred during ResponseConverter#convert");
+            }
+        }
+    }
+
+    /**
+     * Creates a new {@link DynamicHttpFunction} object by combining given {@link DynamicHttpFunction} and
+     * {@link ResponseConverter} objects.
+     */
+    static DynamicHttpFunction of(DynamicHttpFunction function, ResponseConverter converter) {
+        return (ServiceRequestContext ctx, HttpRequest req, Map<String, String> args) -> {
+            Object ret = function.serve(ctx, req, args);
```

Comment:
You should handle the case where `ret` is `instanceof CompletionStage` or `instanceof HttpResponse` here as well.